app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java

```diff
@@ -384,6 +386,119 @@ else if (((selectionEnd <= (_hlEditor.length() - _action.length())) &&
         }
     }
 
+    /**
+     * Set/unset ATX heading level on each selected line
+     *
+     * This routine will make the following conditional changes
+     *
+     * Line is heading of same level as requested -> remove heading
+     * Line is heading of different level that that requested -> add heading of specified level
+     * Line is not heading -> add heading of specified level
+     *
+     * @param level ATX heading level
+     */
+    protected void setHeadingAction(int level) {
+        // Commonmark allows headings of level 1 - 6
+        assert(level >= 1);
+        assert(level <= 6);
+
+        char[] headerChars = new char[level];
+        Arrays.fill(headerChars, '#');
+        String header = new String(headerChars) + ' ';
+
+        ReplacePattern[] patterns = {
+                // Remove extant heading of matching level (preserves leading space)
+                // Commonmark allows up to 3 leading spaces
+                new ReplacePattern(String.format("^(\\s{0,3})#{%d}\\s", level), "$1"),
+                // Convert extant heading to requested level
+                new ReplacePattern("^(\\s{0,3})(#{1,6})\\s", "$1" + header),
+                // Add heading if not a heading
+                new ReplacePattern("^", header),
+        };
+
+        runMarkdownRegexAction(patterns);
+    }
+
+    protected class ReplacePattern {
+        public Pattern search;
+        public String replace;
+        public boolean onlyFirst;
+
+        /**
+         * Construct a ReplacePattern
+         * @param search regex search pattern
+         * @param replace replace string
+         * @param onlyFirst whether to replaceAll or replaceFirst is used
+         */
+        public ReplacePattern(Pattern search, String replace, boolean onlyFirst) {
+            this.search = search;
+            this.replace = replace;
+            this.onlyFirst = onlyFirst;
+        }
+
+        public ReplacePattern(String search, String replace, boolean onlyFirst) {
+            this(Pattern.compile(search), replace, onlyFirst);
+        }
+
+        public ReplacePattern(String search, String replace) {
+            this(Pattern.compile(search), replace);
+        }
+
+        public ReplacePattern(Pattern search, String replace) {
+            this(search, replace, true);
+        }
+    }
+
+    protected void runMarkdownRegexAction(ReplacePattern[] patterns) {
```

Comment:
Anything format specific, i.e. "Markdown" should not be in base class TextActions. Have it in MarkdownTextActions.

However from what I see the bottom most `protected void runMarkdownRegexAction(ReplacePattern[] patterns, boolean matchAll)` is not markdown related. So suggest to rename latter to generic, and move markdown specific stuff to Markdown file(s)

Reply:
Thanks. This was a good suggestion.

I also updated some of the other non-markdown-specific actions on TextActions. (`runMarkdownRegularPrefixAction` -> `runRegularPrefixAction` etc) Let me know if you'd like me to revert this change